query/src/test/java/com/connexta/replication/query/cmp/QueryServiceComponentTest.java

```diff
@@ -0,0 +1,421 @@
+import com.connexta.replication.query.QueryServiceTools;
+import com.connexta.replication.solr.EmbeddedSolrConfig;
+import com.github.npathai.hamcrestopt.OptionalMatchers;
+import io.micrometer.core.instrument.MeterRegistry;
+import java.net.URL;
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.hamcrest.Matchers;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentMatcher;
+import org.mockito.Mockito;
+import org.mockito.stubbing.Answer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.data.solr.repository.config.EnableSolrRepositories;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@ComponentScan(
+    basePackageClasses = {
+      ComponentTestConfig.class,
+      FilterIndexRepository.class,
+      EmbeddedSolrConfig.class
+    })
+@EnableSolrRepositories(basePackages = "com.connexta.replication")
+@RunWith(SpringRunner.class)
+public class QueryServiceComponentTest {
+  private static final int QUEUE_CAPACITY = 3;
+  private static final long THIRTY_MINUTES = TimeUnit.MINUTES.toSeconds(30L);
+  private static final MeterRegistry REGISTRY = new MeterRegistryMock();
+
+  private final QueueBroker broker =
+      new MemoryQueueBroker(
+          QueryServiceComponentTest.QUEUE_CAPACITY, QueryServiceComponentTest.REGISTRY);
+
+  @Autowired private FilterIndexManager indexManager;
+  @Autowired private FilterManager filterManager;
+  @Autowired private SiteManager siteManager;
+
+  @Rule public final MultiThreadedErrorCollector collector = new MultiThreadedErrorCollector();
+
+  private NodeAdapters nodeAdapters = mock(NodeAdapters.class);
+  private NodeAdapterFactory adapterFactory = mock(NodeAdapterFactory.class);
+  private final Set<Task> tasks = new HashSet<>();
+  private final NodeAdapter adapter = mock(NodeAdapter.class);
+
+  @Before
+  public void setup() {
+    when(adapterFactory.create(any(URL.class))).thenReturn(adapter);
+    when(nodeAdapters.factoryFor(SiteType.DDF)).thenReturn(adapterFactory);
+    when(nodeAdapters.factoryFor(SiteType.ION)).thenAnswer(collectAdapterException());
+    when(adapter.isAvailable()).thenReturn(true);
+  }
+
+  private Answer<?> collectAdapterException() {
+    return (Answer<Object>)
+        invocationOnMock -> {
+          Throwable t = new AdapterException("no adapters for ION sites should be retrieved");
+          collector.addError(t);
+          throw t;
+        };
+  }
+
+  @After
+  public void cleanup() throws Exception {
+    tasks.clear();
+    filterManager.objects().map(Filter::getId).forEach(filterManager::remove);
```

Comment:
✏️ It might be worthwhile (in a different ticket) to enhance our embedded solr server such that we can easily ask it to clear the DB which internally would blindly delete the content of all Solr tables. This is much simpler and much more generic than having to rely on our own code to query and delete everything. It is also important that cleaning up aborts testing if it fails to do so otherwise we will end up with leaks in between tests.

We might even be able to autowire that embedded server (or let it be defined here) allowing us to annotate it with `@Rule` such that we can detect the end of a test case and cleanup automatically no matter what.

Reply:
Can we create a ticket for this?